File: app/src/main/java/com/test/murphy/weatherapp/model/WeatherConditions.java
package com.test.murphy.weatherapp.model;

import android.os.Parcel;
import android.os.Parcelable;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Date;

/**
 * Created by wsmurphy on 10/19/17.
 */

public class WeatherConditions implements Parcelable {
    public Double currentTemperature;

    public String currentConditions;
    public String location;
    public String iconId;
    public String iconUrl;

    public Date date;

    public WeatherConditions(JSONObject jsonObject) {
        try {
            //Date object will not exist for current conditions, only for forecast
            Long tempDate = jsonObject.optLong("dt");
            if (tempDate != 0) {
                date = new Date(tempDate * 1000);
            }

            JSONObject mainObject = jsonObject.getJSONObject("main");
            currentTemperature = mainObject.getDouble("temp"); //There is no separate F and C temp in this API

            JSONObject weatherObject = jsonObject.getJSONArray("weather").getJSONObject(0);
            currentConditions = weatherObject.getString("main");

            location = jsonObject.getString("name");

            iconId = weatherObject.getString("icon");
            iconUrl = "http://openweathermap.org/img/w/" + iconId + ".png";
        } catch (JSONException je) {
            //TODO
        } catch (NumberFormatException nfe) {
            //TODO
        }
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public void writeToParcel(Parcel out, int flags) {
        out.writeDouble(currentTemperature);
        out.writeString(currentConditions);
        out.writeString(location);
        out.writeString(iconId);
        out.writeString(iconUrl);
        out.writeLong(date.getTime());
    }

    public static final Parcelable.Creator<WeatherConditions> CREATOR
            = new Creator<WeatherConditions>() {
        public WeatherConditions createFromParcel(Parcel in) {
            return new WeatherConditions(in);
        }

        public WeatherConditions[] newArray(int size) {
            return new WeatherConditions[size];
        }
    };

    private WeatherConditions(Parcel in) {
        currentTemperature = in.readDouble();
        currentConditions = in.readString();
        location = in.readString();
        iconId = in.readString();
        iconUrl = in.readString();
        date = new Date(in.readLong());
    }
}

File: app/src/main/java/com/test/murphy/weatherapp/model/WeatherConditions.kt
package com.test.murphy.weatherapp.model

import android.os.Parcel
import android.os.Parcelable
import org.json.JSONException
import org.json.JSONObject
import java.util.*



/**
 * Created by wsmurphy on 10/19/17.
 */

class WeatherConditions() : Parcelable {
    var currentTemperatureF: Double? = null
    var currentTemperatureC: Double? = null

    var conditionCode: Int = 0
    var currentConditions: String? = null
    var location: String? = null
    var iconId: String? = null
    val iconUrl: String
        get() = "http://openweathermap.org/img/w/$iconId.png"

    var date: Date? = null

    constructor(parcel: Parcel) : this() {
        currentTemperatureF = parcel.readDouble()
        currentTemperatureC = parcel.readDouble()
        conditionCode = parcel.readInt()
        currentConditions = parcel.readString()
        location = parcel.readString()
        iconId = parcel.readString()
    }

    constructor(jsonObject: JSONObject) : this() {
        try {
            //Date object will not exist for current conditions, only for forecast
            val tempDate = jsonObject.optLong("dt")
            if (tempDate != 0L) {
                date = Date(tempDate * 1000)
            }

            val mainObject = jsonObject.getJSONObject("main")

            //There is no separate F and C temp in this API
            //We request F and convert to C so that we don't need to make a second network call when the user switches the units
            currentTemperatureF = mainObject.getDouble("temp")
            currentTemperatureF?.let {
                currentTemperatureC = ftoc(it)
            }

            val weatherObject = jsonObject.getJSONArray("weather").getJSONObject(0)
            currentConditions = weatherObject.getString("main")
            conditionCode = weatherObject.getInt("id")

            location = jsonObject.getString("name")

            iconId = weatherObject.getString("icon")

        } catch (je: JSONException) {
            //TODO
        } catch (nfe: NumberFormatException) {
            //TODO
        }
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeValue(currentTemperatureF)
        parcel.writeValue(currentTemperatureC)
        parcel.writeInt(conditionCode)
        parcel.writeString(currentConditions)
        parcel.writeString(location)
        parcel.writeString(iconId)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<WeatherConditions> {
        override fun createFromParcel(parcel: Parcel): WeatherConditions {
            return WeatherConditions(parcel)
        }

        override fun newArray(size: Int): Array<WeatherConditions?> {
            return arrayOfNulls(size)
        }
    }

    //Convert fahrenheit to celsius
    private fun ftoc(f: Double): Double {
        return f - 32 / 1.8
    }
}
